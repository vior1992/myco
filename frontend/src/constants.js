const PROPS_TO_DISPLAY = [
  { key: 'first_name', title: 'First name' },
  { key: 'last_name', title: 'Last name' },
  { key: 'email', title: 'Email' },
  { key: 'phone_number', title: 'Phone number' },
];

const MOCKED_CONTACTS = [
  {
    id: 1,
    first_name: 'Ragnar',
    last_name: 'Lodbrok',
    email: 'vikings@mocked.com',
    phone_number: '+ 00 00 01',
  },
  {
    id: 2,
    first_name: 'Bj√∂rn',
    last_name: 'Ironside',
    email: 'vikings@mocked.com',
    phone_number: '+ 00 00 02',
  },
];

export { PROPS_TO_DISPLAY, MOCKED_CONTACTS };
