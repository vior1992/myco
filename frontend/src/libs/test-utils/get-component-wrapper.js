import React from 'react';
import { ToastProvider } from 'react-toast-notifications';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { GlobalStyles, Theme } from '../../styles';

const getComponentWrapper = ({
  Component,
  mockedProps = {},
  initialRoute = '/',
}) => {
  const history = createMemoryHistory({ initialEntries: [initialRoute] });

  return (
    <Theme>
      <GlobalStyles />
      <ToastProvider>
        <Router history={history}>
          <Component {...mockedProps} />
        </Router>
      </ToastProvider>
    </Theme>
  );
};

export default getComponentWrapper;
