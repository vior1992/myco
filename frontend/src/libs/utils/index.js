import buildRequestOptions from './build-request-options';

export { buildRequestOptions };
