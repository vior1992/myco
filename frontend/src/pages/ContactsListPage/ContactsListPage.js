import React, { useEffect, useState } from 'react';
import { useToasts } from 'react-toast-notifications';
import {
  StyledContactsListPageWrapper,
  StyledTitlesWrapper,
  StyledLink,
  StyledContactList,
} from './StyledContactsListPage';
import { ContactCard } from '../../components';
import { contactsRepository } from '../../repositories';

const ContactsListPage = () => {
  const [contacts, setContacts] = useState([]);
  const [hasToRefetch, setHasToRefetch] = useState(false);
  const { addToast } = useToasts();

  useEffect(() => {
    const fetchData = async () => {
      const data = await contactsRepository.getContacts({ addToast });

      setContacts(data);
      setHasToRefetch(false);
    };
    fetchData();
  }, [hasToRefetch, addToast]);

  const handleDeleteContact = async (id) => {
    await contactsRepository.deleteContact({
      id,
      addToast,
    });

    setHasToRefetch(true);
  };

  return (
    <StyledContactsListPageWrapper>
      <StyledTitlesWrapper>
        <h1>
          MYCO
          <span role="img" aria-label="monkey">
            🐒
          </span>
        </h1>
        <h2>My Contact List</h2>
      </StyledTitlesWrapper>
      <StyledLink to="/contacts">
        <i className="fas fa-user-plus"></i>
      </StyledLink>
      <StyledContactList>
        {contacts.map((contact) => (
          <ContactCard
            key={contact.id}
            {...contact}
            onClickDeleteContact={handleDeleteContact}
          />
        ))}
      </StyledContactList>
    </StyledContactsListPageWrapper>
  );
};

export default ContactsListPage;
