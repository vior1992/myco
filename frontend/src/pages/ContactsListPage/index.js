import ContactsListPage from './ContactsListPage';

export default ContactsListPage;
