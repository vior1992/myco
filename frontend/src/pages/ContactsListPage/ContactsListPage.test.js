import { render, act, screen, fireEvent } from '@testing-library/react';
import ContactsListPage from './ContactsListPage';
import { MOCKED_CONTACTS } from '../../constants';
import { getComponentWrapper } from '../../libs/test-utils';

const setup = () => {
  const wrappedComponent = getComponentWrapper({ Component: ContactsListPage });
  const _render = render(wrappedComponent);
  return { ..._render };
};

describe('ContactsListPage component', () => {
  beforeAll(() => jest.spyOn(window, 'fetch'));

  beforeEach(() => {
    fetch.mockClear();
    window.fetch.mockResolvedValueOnce({
      ok: true,
      json: async () => MOCKED_CONTACTS,
    });
  });

  it('should render without throwing an error & match snapshot', async () => {
    await act(async () => {
      const { asFragment } = setup();
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it('should delete contact on click delete icon', async () => {
    await act(async () => {
      setup();
    });

    const [deleteIcon] = screen.getAllByTestId('delete-icon');
    fireEvent.click(deleteIcon);

    expect(window.fetch).toHaveBeenCalledWith(
      'http://localhost:3001/api/contacts/1',
      expect.objectContaining({
        method: 'DELETE',
      })
    );
  });
});
