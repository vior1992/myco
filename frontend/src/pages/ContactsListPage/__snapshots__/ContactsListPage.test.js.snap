// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactsListPage component should render without throwing an error & match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnylx hRUlCE"
  >
    <div
      class="sc-gtssRu bLovyb"
    >
      <h1>
        MYCO
        <span
          aria-label="monkey"
          role="img"
        >
          üêí
        </span>
      </h1>
      <h2>
        My Contact List
      </h2>
    </div>
    <a
      class="sc-dlnjPT hGMcqv"
      href="/contacts"
    >
      <i
        class="fas fa-user-plus"
      />
    </a>
    <div
      class="sc-hKFyIo jetHVG"
    />
  </div>
</DocumentFragment>
`;
